doc/man3/EVP_PKEY_sign.pod

```diff
@@ -26,37 +35,121 @@ for more information about implicit fetches.
 sets the passed parameters I<params> on the context before returning.
 
-The EVP_PKEY_sign() function performs a public key signing operation
-using I<ctx>. The data to be signed is specified using the I<tbs> and
-I<tbslen> parameters. If I<sig> is NULL then the maximum size of the output
-buffer is written to the I<siglen> parameter. If I<sig> is not NULL then
-before the call the I<siglen> parameter should contain the length of the
-I<sig> buffer, if the call is successful the signature is written to
-I<sig> and the amount of data written to I<siglen>.
+EVP_PKEY_sign_init_ex2() initializes a public key algorithm context I<ctx> for
+signing a pre-computed message digest using the algorithm given by I<algo> and
+the key given through L<EVP_PKEY_CTX_new(3)> or L<EVP_PKEY_CTX_new_from_pkey(3)>.
+A context I<ctx> without a pre-loaded key cannot be used with this function.
+This function provides almost the same functionality as EVP_PKEY_sign_init_ex(),
+but is uniquely intended to be used with a pre-computed messsage digest, and
+allows pre-determining the exact conditions for that message digest, if a
+composite signature algorithm (such as RSA-SHA256) was fetched.
+Following a call to this function, setting parameters that modifies the digest
+implementation or padding is not normally supported.
+
+EVP_PKEY_sign_message_init() initializes a public key algorithm context I<ctx>
+for signing an unlimited size message using the algorithm given by I<algo> and
+the key given through L<EVP_PKEY_CTX_new(3)> or L<EVP_PKEY_CTX_new_from_pkey(3)>.
+Passing the message is supported both in a one-shot fashion using
+EVP_PKEY_sign(), and through the update fashion functions
+EVP_PKEY_sign_message_update() and EVP_PKEY_sign_message_final().
+This function enables using algorithms that can process infinite input, such
+as ED25519, RSA-SHA256 and similar.
+
+EVP_PKEY_sign_message_update() adds I<inlen> bytes from I<in> to the data to be
+processed for signature.  The signature algorithm specification and
+implementation determine how the input bytes are processed and if there's a
+limit on the total size of the input.  See L</NOTES> below for a deeper
+explanation.
+
+EVP_PKEY_sign_message_final() signs the processed data and places the data in
+I<sig>, and the number of signature bytes in I<*siglen>, if the number of
+bytes doesn't surpass the size given by I<sigsize>.
+I<sig> may be NULL, and in that case, only I<*siglen> is updated with the
+number of signature bytes.
+
+EVP_PKEY_sign() is a one-shot function that can be used with all the init
+functions above.
+When initialization was done with EVP_PKEY_sign_init(), EVP_PKEY_sign_init_ex()
+of EVP_PKEY_sign_init_ex2(), the data specified by I<tbs> and I<tbslen> is
+signed after appropriate padding.
+When initialization was done with EVP_PKEY_sign_message_init(), the data
+specified by I<tbs> and I<tbslen> is digested by the implied message digest
```

Comment:
tbs = to be signed (Is this an appropriate name now? - is data any better as a name)

Reply:
Considering that *tbs* is a term already used with existing functionality, and is also a term used elsewhere (see X.509 with `tbsCertlist` and `tbsCertificate`), I think our use of *tbs* is appropriate here.